ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java

```diff
@@ -0,0 +1,158 @@
+
+import avro.shaded.com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.util.KsqlException;
+
+/**
+ * This class takes a Method that has been marked with the Udf annotation,
+ * generates code for it to be invoked and compiles it into an UdfInvoker instance.
+ */
+public class UdfCompiler {
+  private static final Logger logger = LoggerFactory.getLogger(UdfCompiler.class);
+
+  private static final Map<Class, Function<Integer, String>> typeConverters
+      = ImmutableMap.<Class, Function<Integer, String>>builder()
+      .put(int.class, index -> typeConversionCode("Integer", index))
+      .put(Integer.class, index -> typeConversionCode("Integer", index))
+      .put(long.class, index -> typeConversionCode("Long", index))
+      .put(Long.class, index -> typeConversionCode("Long", index))
+      .put(double.class, index -> typeConversionCode("Double", index))
+      .put(Double.class, index -> typeConversionCode("Double", index))
+      .put(boolean.class, index -> typeConversionCode("Boolean", index))
+      .put(Boolean.class, index -> typeConversionCode("Boolean", index))
+      .put(String.class, index -> typeConversionCode("String", index))
+      .put(Map.class, index -> typeConversionCode("Map", index))
+      .put(List.class, index -> typeConversionCode("List", index))
+      .build();
+
+
+  UdfInvoker compile(final Method method, final ClassLoader loader) {
+    try {
+      final IScriptEvaluator scriptEvaluator = createScriptEvaluator(method, loader);
+      final String code = generateCode(method);
+      return (UdfInvoker) scriptEvaluator.createFastEvaluator(code,
+          UdfInvoker.class, new String[]{"thiz", "args"});
+    } catch (final KsqlException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new KsqlException("Failed to compile udf wrapper class for "
+          + method, e);
+    }
+  }
+
+  /**
+   * Generates code for the given method. Assuming the method has a single Boolean argument
+   * the generated code would look like the below. A block of code like this is generated for each
+   * of the arguments.
+   *<pre>{@code
+   * // try and coerce the arguments to the types expected by the UDF
+   * Boolean arg0;
+   * if(args[0] == null) arg0 = null;
+   * else if(args[0] instanceof Boolean) arg0 = (Boolean)args[0];
+   * else if(args[0] instanceof String) arg0 = Boolean.valueOf((String)args[0]);
+   * else throw new KsqlFunctionException("Type: " + args[0].getClass()
+   * + "is not supported by KSQL UDFS");
+   *
+   * // invoke the udf with the args
+   * return ((UdfCompilerTest) thiz).udf(arg0);
+   *}</pre>
+   * @param method  the UDF to generate the code for
+   * @return String representation of the code that should be compiled for the UDF
+   */
+
+  private static String generateCode(final Method method) {
+    final StringBuilder builder = new StringBuilder();
+    final Class<?>[] params = method.getParameterTypes();
+    for (int i = 0; i < params.length; i++) {
```

Comment:
Use `IntString.range(0, params.length) ... collect(Collectors.joining(\n))` to avoid StringBuilder.

Reply:
Thoughts?